tools/ioflow/ioflow-syncwrite.c

```diff
@@ -0,0 +1,241 @@
+        valp->ts_writepg = bpf_ktime_get_ns();
+    return 0;
+}
+
+int ext4_sync_entry(struct pt_regs *ctx) {
+    u64 pid =  bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp && valp->ts_ext4sync == 0)
+        valp->ts_ext4sync = bpf_ktime_get_ns();
+    return 0;
+}
+
+/* BLock layer */
+int block_entry(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_block_entry(valp, ts);
+    }
+    return 0;
+}
+
+int block_return(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_block_return(valp, ts);
+    }
+    return 0;
+}
+
+int split_entry(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_split_entry(valp, ts);
+    }
+    return 0;
+}
+
+int split_return(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_split_return(valp, ts);
+    }
+    return 0;
+}
+
+int merge_entry(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_merge_entry(valp, ts);
+    }
+    return 0;
+}
+
+int merge_return(struct pt_regs *ctx) {
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        comm_merge_return(valp, ts);
+    }
+    return 0;
+}
+
+// Async request handling
+int rq_create(struct pt_regs *ctx, struct request *rq) {
+    // Still in the syscall process's context now. 
+    u64 ts = bpf_ktime_get_ns();
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = syscall_map.lookup(&pid);
+    if (valp) {
+        struct rqval_t rqval = {0};
+        comm_rq_create(&rqval, valp, rq, pid >> 32, ts);
+        request_map.insert(&rq, &rqval);
+    }
+    return 0;
+}
+
+// The request is issued to device driver.
+int rq_issue(struct pt_regs *ctx, struct request *rq) {
+    // Async to the syscall process now. 
+    u64 ts = bpf_ktime_get_ns();
+    struct rqval_t *rqvalp = request_map.lookup(&rq);
+    comm_rq_issue(rqvalp, ts);
+    return 0;
+}
+
+// The request is done.
+int rq_done(struct pt_regs *ctx, struct request *rq) {
+    u64 ts = bpf_ktime_get_ns();
+    struct rqval_t *rqvalp = request_map.lookup(&rq);
+    if (rqvalp) {
+        struct rqdata_t rqdata = {0};
+        comm_rq_done(&rqdata, rqvalp, ts);
+
+        // filter out data that has latency less than the threshold
+        if (rqdata.queue + rqdata.service >= REQUEST_THRESHOLD) {
```

Comment:
hmmm, so this means we won't get all the requests for a particular syscall, could it be possible to have a single threshold, if the vfs took longer than X, then output all requests associated with it?

Reply:
Well it may be kinda hard, bc that may require dynamic storage (and allocation) of bpf map/list entries for cached request records in kernel space. And we have to loop through the bpf data structure when emitting the request records. Maybe doable if BCC/eBPF provides linked list data structure in kernel space?